github/messages.go

```diff
@@ -132,17 +137,37 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 //       // Process payload...
 //     }
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
```

Comment:
Since "application/x-www-form-urlencoded" and "application/json" are the only two content types we expect from GitHub, and support, let's explicitly check it's one of those two and return an error otherwise. It wouldn't make sense to treat it as JSON if the content type is not actually "application/json".

This is a good place to use a `switch`, like this:

```Go
switch r.Header.Get("Content-Type") {
case "application/x-www-form-urlencoded":
    ...
case "application/json":
    ...
default:
    return nil, fmt.Errorf("Webhook payload has unsupported Content-Type %q", r.Header.Get("Content-Type"))
}
```